core/config/config_helper.hpp

```diff
@@ -0,0 +1,220 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_CORE_CONFIG_CONFIG_HELPER_HPP_
+#define GKO_CORE_CONFIG_CONFIG_HELPER_HPP_
+
+
+#include <string>
+#include <type_traits>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/config/config.hpp>
+#include <ginkgo/core/config/registry.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+namespace gko {
+namespace config {
+
+
+/**
+ * LinOpFactoryType enum is to avoid forward declartion, linopfactory header,
+ * two template versions of build_from_config
+ */
+enum class LinOpFactoryType : int { Cg = 0 };
+
+
+/**
+ * It is only an intermediate step after dispatching the class base type. Each
+ * implementation needs to deal with the template selection.
+ */
+template <LinOpFactoryType flag>
+deferred_factory_parameter<gko::LinOpFactory> build_from_config(
+    const pnode& config, const registry& context,
+    type_descriptor td = type_descriptor{"void", "void"});
+
+/**
+ * This function is to update the default type setting from current config.
+ *
+ * @note It might update the unused type for the current class.
+ */
+type_descriptor update_type(const pnode& config, const type_descriptor& td);
+
+/**
+ * get_pointer searches the pointer in the registry by string
+ */
+template <typename T>
+inline std::shared_ptr<T> get_pointer(const pnode& config,
```

Comment:
This name (and for the other `get_*` function) is not suitable. `get` is just too non-descriptive. I could imagine `parse_*` for these functions. Then the connection to the top-level function would also be clearer.

Also maybe call it `*_custom_data`, because this will only give the custom data stored in the registry.

Reply:
No, the `get_pointer` can get the stored objected or generate from the config directly.

Reply:
So it's doing two things, either create a new object, or retrieve an existing one. Maybe split this into two functions.

Reply:
it is already splitted into the two function. There's no necessary to increase the difficulty on parse again

Reply:
Wait, can we go back? From the implementation, this clearly doesn't create/generate a new object, it only retrieves an object from the registry.

Reply:
ah, you are right. I mean get_factory can be stored object or from config

Reply:
yes, that makes sense. And here I would say that `pointer` is not descriptive enough, since we know that it will return a user provided object.

Reply:
I use `get_stored_obj` now

Reply:
Thanks, that sounds good!